### ISSUE TITLE
Using Latex for equations in titles or labels

### ISSUE THREAD

**Issue by PaulWessel** [GenericMappingTools/gmt#4563]
**Description of the desired feature**

This issue comes up from time to time: Why can't I use Latex to specify an equation that I want to be an axis label or part of a title in GMT?   **Answer**: We have no such implementation.

I actually was going down this road a few years ago but got side-tracked and/or discouraged.  Decided to have another look.  Let's say I want to place this (meaningless) equation in my y-axis label:

`Solution to $\nabla^4 \Psi = \frac{\delta w}{\delta t} = f(x(t))$`

If a label or title (or custom annotation item, I suppose) were recognized to contain Latex (is the $ enough? Or would we need @$ ...@$ to flag that in GMT?), then we could build a tiny Latex file (say _eq.tex_ here):

```
\documentclass{article}
\begin{document}
\thispagestyle{empty}
Solution to $\nabla^4 \Psi = \frac{\delta w}{\delta t} = f(x(t))$
\end{document}
```

Now, GMT would have to call **latex** and **dvips**, which would be new run-time requirements if you wish to use Latex in GMT (plus give errors if you try this without **latex** installed, like for **gs** or **ffmpeg**).  Doing this by hand, I ran `latex eq` which produces _eq.dvi_.  I can now call `dvips -E eq.dvi -o eq.eps `to get the EPS.  Can this file be placed in GMT? Well, I can run `gmt image eq.eps -Dx2i/2i+w0.5i -png eq `and get

![eq](https://user-images.githubusercontent.com/26473567/101848252-1778ea00-3af9-11eb-80aa-797411272627.png)

so yes, it can.  The scheme would then be to place this EPS file centered on the balance point we use to center the title or label.  So in principle this should work.  Some minor issues or points to make:

1. This would need to be done via system calls as I do not see any other way.  Presumably, the two commands can be written into a script (*.sh or *.bat) and run via system.  GMT would then place the EPS file where it needs to go.
2. There may be a need to set specific font sizes since the above is whatever equations are in a latex document (12p?) and we may want different sizes.  So need to set the right font size.Latex gurus can advice on this.
3. I am not sure about font choices.  If **LABEL_FONT** is Helvetica, presumably we would want the "Solution to" part to be Helvetica?

Comments welcome, @GenericMappingTools/core and @GenericMappingTools/gmt-contributors.

**Comment by meghanrjones** [comment]
My preference would be to create a 'how-to-guide' for people to put nice LaTex-produced raster files in their GMT plots. Just seems like it could be a slippery slope with people asking for more and more preamble options. Open to being convinced otherwise though...

**Comment by PaulWessel** [comment]
Fair comment.  Usually, if the developer(s) find the feature useful for themselves or in the support of the scientists that the funders expect us to support, and it is not _too_ hard, then we tend to add it.  Given the above I think this is a few hours for me to implement and bring me joy in the process(!).  Like comfort food.

**Comment by seisman** [comment]
> My preference would be to create a 'how-to-guide' for people to put nice LaTex-produced raster files in their GMT plots.

I agree with @meghanrjones. An example should be enough.

**Comment by joa-quim** [comment]
Just a way to include **whatever** via call to psimage would be quite nice. It could be an PS equation, an image, etc...

**Comment by PaulWessel** [comment]
Not sure I understand.  image can plot EPS and rasters.

**Comment by joa-quim** [comment]
Well, we were talking about inserting Latex EPS in titles and legends. What I meant was, if we had a generic way of calling ``psimage``, we could insert anything ``psimage`` can deal with.

**Comment by PaulWessel** [comment]
Do you mean stuff like (made up syntax)

`-B+t=eq.eps`

and have GMT deal with proper placement?

**Comment by joa-quim** [comment]
Yes, sort off. This is the kind of thing you would have to do in your harmonic example..

**Comment by PaulWessel** [comment]
Does MATLAB, matplotlib, etc, let you place images as part of annotations and titles, or do they let you supply Latex equations?  I think the latter only, no?  Seems odd to me that we want to block what people are used to and make them do a separate dance to create the EPS when this seems trivial for us to implement. Does not sound like a selling point for GMT to me.

**Comment by meghanrjones** [comment]
I am a big fan of comfort food (and equations), so fine by me if Paul wants to move forward.

**Comment by joa-quim** [comment]
I don't think they can place images, but if we could that would be a selling point for us. The fact that Titles/Labels could accept *image* input does not invalidate your what you want to do. In fact from my understanding you will have to implement most of what is need to accept also other things (images). 

Latex is gigantic and perhaps alternatives like https://www.mathjax.org/ (which I think only works for browsers) could be used as well to create equations.

**Comment by PaulWessel** [comment]
Yes, latex is big (took a while to install it again on my mac), but presumably anyone using Latex to write papers and is knowledgable about latex already has latex installed.  Anyway mathjax offers an API but has no documentation so not clear what it is and if it would allow calls to something useful.

**Comment by joa-quim** [comment]
I've been trying to say that there seem to exist alternatives. Don't know how it works but that ``GR`` package that I linked a couple times is able to do this without any Latex instaled

![image](https://user-images.githubusercontent.com/537321/101961951-c9500d80-3c02-11eb-8203-ae3b51491b00.png)

**Comment by GD-A** [comment]
> Yes, latex is big (took a while to install it again on my mac), but presumably anyone using Latex to write papers and is knowledgable about latex already has latex installed. 

Wouldn't it be easier to simply adopt Latex syntax (GMT 7?) instead of relying on a latex distro ?
-B+t can become -B+\title{} ...

As for the core problem itself, Latex "bare" is not that big, the million modules are... But presumably GMT would need only a bare version + some maths modules ...

**Comment by PaulWessel** [comment]
My PR that implements Latex in **-B** will be submitted by weekend's end.  Just testing a few things.  Already working for titles but need to add axes labels.

**Comment by uleysky** [comment]
Some potential issues when incorporating LaTeX formulas into text (@$ or something like):
1. Baseline alignment. Is it possible in principle to determine the location of the baseline in a arbitrary formula in order to correctly align its insertion into the text? Try to insert some continued fraction.
2. Matching font sizes. Most likely, this can be solved by selecting the stretch/compression ratio of the inserted image.
3. Matching font families. Difficult, if not impossible.

On the other hand, inserting a formula as a separate object does not create these problems and would be very convenient. I would suggest adding a separate text rendering mode via LaTeX. That is, the string is rendered either by the GMT or by LaTeX, without mixed mode. And, of course, the user should be able to define his own LaTeX template, because the LaTeX has many packages.

**Comment by PaulWessel** [comment]
You can see #4566 to see if you like it.  It does address 2,3 and probably 1 since I typeset the entire text in Latex if it contains an equation, which is what your final point is about.  I am aiming at simplifying this for 99% of the cases but there is always outlier cases.  If it truly becomes necessary to allow adding more Latex packages etc then we can see.
